Mathlib/Analysis/InnerProductSpace/Orientation.lean

```diff
@@ -0,0 +1,381 @@
+    o.volumeForm = -b.toBasis.det := by
+  cases' n with n
+  · classical
+    have : positiveOrientation ≠ o := by rwa [b.toBasis.orientation_isEmpty] at hb
+    simp [volumeForm, Or.by_cases, dif_neg this.symm]
+    tauto
+  let e : OrthonormalBasis (Fin n.succ) ℝ E := o.finOrthonormalBasis n.succ_pos Fact.out
+  simp [volumeForm]
+  apply e.det_eq_neg_det_of_opposite_orientation b
+  convert hb.symm
+  exact o.finOrthonormalBasis_orientation _ _
+#align orientation.volume_form_robust_neg Orientation.volumeForm_robust_neg
+
+@[simp]
+theorem volumeForm_neg_orientation : (-o).volumeForm = -o.volumeForm := by
+  cases' n with n
+  · refine' o.eq_or_eq_neg_of_isEmpty.elim _ _
+    · rintro rfl
+      simp [volumeForm_zero_neg]
+    · rintro rfl
+      sorry
+  let e : OrthonormalBasis (Fin n.succ) ℝ E := o.finOrthonormalBasis n.succ_pos Fact.out
+  have h₁ : e.toBasis.orientation = o := o.finOrthonormalBasis_orientation _ _
+  have h₂ : e.toBasis.orientation ≠ -o := by
+    symm
+    rw [e.toBasis.orientation_ne_iff_eq_neg, h₁]
+  rw [o.volumeForm_robust e h₁, (-o).volumeForm_robust_neg e h₂]
+#align orientation.volume_form_neg_orientation Orientation.volumeForm_neg_orientation
+
+theorem volumeForm_robust' (b : OrthonormalBasis (Fin n) ℝ E) (v : Fin n → E) :
+    |o.volumeForm v| = |b.toBasis.det v| := by
+  cases n
+  · refine' o.eq_or_eq_neg_of_isEmpty.elim _ _ <;> rintro rfl <;> simp
+  ·
+    rw [o.volumeForm_robust (b.adjustToOrientation o) (b.orientation_adjustToOrientation o),
+      b.abs_det_adjustToOrientation]
+#align orientation.volume_form_robust' Orientation.volumeForm_robust'
+
+/-- Let `v` be an indexed family of `n` vectors in an oriented `n`-dimensional real inner
+product space `E`. The output of the volume form of `E` when evaluated on `v` is bounded in absolute
+value by the product of the norms of the vectors `v i`. -/
+theorem abs_volumeForm_apply_le (v : Fin n → E) : |o.volumeForm v| ≤ ∏ i : Fin n, ‖v i‖ := by
+  cases' n with n
+  · refine' o.eq_or_eq_neg_of_isEmpty.elim _ _
+    · rintro rfl
+      simp only [Nat.zero_eq, volumeForm_zero_pos, AlternatingMap.constLinearEquivOfIsEmpty_apply,
+        AlternatingMap.constOfIsEmpty_apply, Function.const_apply, abs_one, Finset.univ_eq_empty, Finset.prod_empty,
+        le_refl]
+    · rintro rfl
+      simp only [Nat.zero_eq, volumeForm_neg_orientation, volumeForm_zero_pos,
+        AlternatingMap.constLinearEquivOfIsEmpty_apply, AlternatingMap.neg_apply, AlternatingMap.constOfIsEmpty_apply,
+        Function.const_apply, abs_neg, abs_one, Finset.univ_eq_empty, Finset.prod_empty, le_refl]
+  haveI : FiniteDimensional ℝ E := fact_finiteDimensional_of_finrank_eq_succ n
+  have : finrank ℝ E = Fintype.card (Fin n.succ) := by simpa using _i.out
+  let b : OrthonormalBasis (Fin n.succ) ℝ E := gramSchmidtOrthonormalBasis this v
+  have hb : b.toBasis.det v = ∏ i, ⟪b i, v i⟫ := gramSchmidtOrthonormalBasis_det this v
```

Comment:
So this is fun. `volumeForm_robust'` is a statement about `@Basis.det .. (fun a b ↦ instDecidableEqFin n a b)` (https://leanprover-community.github.io/mathlib4_docs/Init/Prelude.html#instDecidableEqFin), while `gramSchmidtOrthonormalBasis_det` is a statement about `@Basis.det  .. (fun a b ↦ instDecidableEq a b)` (https://leanprover-community.github.io/mathlib4_docs/Mathlib/Init/Algebra/Order.html#instDecidableEq)